test/integration/apiserver/flowcontrol/concurrency_test.go

```diff
@@ -0,0 +1,249 @@
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http/httptest"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/prometheus/common/expfmt"
+	"github.com/prometheus/common/model"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	flowcontrolv1alpha1 "k8s.io/api/flowcontrol/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	genericfeatures "k8s.io/apiserver/pkg/features"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+	"k8s.io/kubernetes/pkg/master"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+const (
+	dispatchedRequestCountMetricsName               = "apiserver_flowcontrol_dispatched_requests_total"
+	dispatchedRequestCountMetricsLabelPriorityLevel = "priorityLevel"
+	timeout                                         = time.Second * 10
+)
+
+func setup(t testing.TB) (*httptest.Server, *rest.Config, framework.CloseFunc) {
+	opts := framework.MasterConfigOptions{EtcdOptions: framework.DefaultEtcdOptions()}
+	opts.EtcdOptions.DefaultStorageMediaType = "application/vnd.kubernetes.protobuf"
+	masterConfig := framework.NewIntegrationTestMasterConfigWithOptions(&opts)
+	resourceConfig := master.DefaultAPIResourceConfigSource()
+	resourceConfig.EnableVersions(schema.GroupVersion{
+		Group:   "flowcontrol.apiserver.k8s.io",
+		Version: "v1alpha1",
+	})
+	masterConfig.GenericConfig.MaxRequestsInFlight = 5
+	masterConfig.GenericConfig.MaxMutatingRequestsInFlight = 5
+	masterConfig.GenericConfig.OpenAPIConfig = framework.DefaultOpenAPIConfig()
+	masterConfig.ExtraConfig.APIResourceConfigSource = resourceConfig
+	_, s, closeFn := framework.RunAMaster(masterConfig)
+
+	return s, masterConfig.GenericConfig.LoopbackClientConfig, closeFn
+}
+
+func TestPriorityLevelIsolation(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.APIPriorityAndFairness, true)()
+	// NOTE: disabling the feature should fail the test
+	_, loopbackConfig, closeFn := setup(t)
+	defer closeFn()
+
+	loopbackClient := clientset.NewForConfigOrDie(loopbackConfig)
+	noxu1Client := getClientFor(loopbackConfig, "noxu1")
+	noxu2Client := getClientFor(loopbackConfig, "noxu2")
+
+	priorityLevelNoxu1, _, err := createPriorityLevelAndBindingFlowSchemaForUser(loopbackClient, "noxu1")
+	require.NoError(t, err)
+	priorityLevelNoxu2, _, err := createPriorityLevelAndBindingFlowSchemaForUser(loopbackClient, "noxu2")
+	require.NoError(t, err)
+
+	wg := &sync.WaitGroup{}
+	// "elephant"
+	streamRequests(wg, 10, 100, func() {
+		_, err := noxu1Client.CoreV1().Namespaces().List(context.TODO(), metav1.ListOptions{})
+		require.NoError(t, err)
+	})
+
+	streamRequests(nil, 1, 100, func() {
+		_, err := noxu2Client.CoreV1().Namespaces().List(context.TODO(), metav1.ListOptions{})
```

Comment:
Is this the most time-consuming request we can come up with?  Longer-running requests give a better signal-to-noise ratio.  Although, as discussed elsewhere, this test does not look closely enough at the behavior for this to matter.

Reply:
i was looking for somewhere to inject such a filter. but it doesnt seem to work for an integration-testing apiserver..

Reply:
If you really wanted to be fancy, you could register a webhook. It might be worth it if this is flaky.

Reply:
(The webhook could be used to force some sort of logical clock into the system.)

Reply:
@yue9944882 : right, but your reply is not attached to the correct original comment.  But I guess we are discussing it here now.  So...

In the approach I suggested at https://github.com/kubernetes/kubernetes/pull/89256#issuecomment-603658953 , there is no apiserver.  We just write a test that instantiates the APF filter and exercises it to prove that it separates traffic.  Proving the integration of the filter into the server is separated into a distinct test.  The code here can be simplified to become that test.  We only need to have the client make a few requests and check that they got dispatched.

@lavalamp: yes, we could use a real apiserver and add a server for a webhook and configure its use and have a coordination mechanism between the test func and that webhook server.  But that is a lot more complicated than necessary.  I think we can test the behavior of the filter directly, and it will be much simpler.

Reply:
> I think we can test the behavior of the filter directly

I agree, however, I think there is also value in a test like this that ensures everything has been wired correctly.

Reply:
That's what I meant about "proving the integration of the filter".  Yes, we need a test that everything is wired correctly.  A simplified verison of this PR will do that.  We only need it to introduce one distinct PL & FS, submit a few requests, and check that they got dispatched.